src/EndSwitches.cpp

```diff
@@ -0,0 +1,202 @@
+            {
+                stopSlewDir |= TRACKING;
+            }
+
+            // Set state first thing to avoid re-entrancy in call to waitUntilStopped() below.
+            _state = _state == EndSwitchState::SWITCH_AT_MINIMUM ? EndSwitchState::WAIT_FOR_STOP_AT_MINIMUM
+                                                                 : EndSwitchState::WAIT_FOR_STOP_AT_MAXIMUM;
+            LOG(DEBUG_MOUNT,
+                "[ENDSWITCH]: Slewing is active, so stopping slew on %s axis (%x, %x). State is WAIT_FOR_STOP",
+                _axis == StepperAxis::RA_STEPS ? "RA" : "DEC",
+                stopSlewDir,
+                _dir);
+            _pMount->stopSlewing(stopSlewDir);
+            _pMount->waitUntilStopped(_dir);  // This function will call checkSwitchState
+            LOG(DEBUG_MOUNT,
+                "[ENDSWITCH]: Slewing stopped on %s axis. (%x, %x)",
+                _axis == StepperAxis::RA_STEPS ? "RA" : "DEC",
+                _pMount->slewStatus(),
+                _pMount->mountStatus());
+            long currentPos       = _pMount->getCurrentStepperPosition(_dir);
+            long backDistance     = _posWhenTriggered - currentPos;
+            long backSlewDistance = (12 * backDistance) / 10;  // Go back 120% distance that we ran past the switch
+            LOG(DEBUG_MOUNT,
+                "[ENDSWITCH]: Reached maximum at %l. Stopped at %l (%l delta), moving back 1.2x (%l). Stopped tracking (if RA at Max). "
+                "State is SWITCH_SLEWING_OFF",
+                _posWhenTriggered,
+                currentPos,
+                backDistance,
+                backSlewDistance);
+            _pMount->moveStepperBy(_axis, backSlewDistance);
+            _state = _state == EndSwitchState::WAIT_FOR_STOP_AT_MINIMUM ? EndSwitchState::SWITCH_SLEWING_OFF_MINIMUM
+                                                                        : EndSwitchState::SWITCH_SLEWING_OFF_MAXIMUM;
+        }
+        else if ((_state == EndSwitchState::SWITCH_AT_MAXIMUM) && _pMount->isSlewingTRK())
+        {
+            // If we are not slewing, we must have tracked into the maximum, so just stop tracking and wait for it to be moved off maximum manually.
+            LOG(DEBUG_MOUNT,
+                "[ENDSWITCH]: Reached maximum position on %s axis while tracking. Stopping tracking. State is now SWITCH_SLEWING_OFF",
+                _axis == StepperAxis::RA_STEPS ? "RA" : "DEC");
+            _pMount->stopSlewing(TRACKING);
+            _state = EndSwitchState::SWITCH_SLEWING_OFF_MAXIMUM;
+        }
+        else if (_state == EndSwitchState::SWITCH_AT_MINIMUM)
+        {
+            // This shouldn't really happen, since we can't really reach minimum if we're not slewing, since tracking goes to maximum.
+            // But just so that the state machine resets correctly once we move off this switch we transition to the state that waits for that.
+            _state = EndSwitchState::SWITCH_SLEWING_OFF_MINIMUM;
```

Comment:
If this is an invalid state, adding a log would be helpful here

Reply:
Yeah.